### ISSUE TITLE
(Maybe bug?) optional fields arent optional

### ISSUE THREAD

**Issue by MFogleman** [jquense/yup#1158]
I'm creating an endpoint that takes a payload that requires some fields, but others are optional.  The optional fields can be missing.  I am using yup for the handler validation, and the type definition, so that the 2 are coupled.  This pattern was previously working as of 0.29.3, but i'm not able to recreate this behavior in 0.32.6

Marking a field as optional in a schema, and inferring the type, does not make it optional on the type, it makes it a union of `TheType | undefined`

Example in the sandbox of it making TS mad:
https://codesandbox.io/s/zen-wu-4uihp?file=/src/index.test.ts

Heres the code from the sandbox, for convenience:
```typescript
import type { Asserts, TypeOf } from "yup";
import * as yup from "yup";

const inputSchema = yup.object({
  age: yup.number().defined(),
  favNum: yup.number().optional()
});

interface InputViaTypeOf extends TypeOf<typeof inputSchema> {}

const input1: InputViaTypeOf = { age: 4 };
// Property 'favNum' is missing in type '{ age: number; }' but required in type 'InputViaTypeOf'.ts(2741)

const input2: InputViaTypeOf = { age: 4, favNum: 7 }; // works fine
const input3: InputViaTypeOf = { age: 4, favNum: undefined }; // works fine

/**
 * Trying it with extending Asserts
 */
interface InputViaAssert extends Asserts<typeof inputSchema> {}

const input4: InputViaAssert = { age: 4 };
// Property 'favNum' is missing in type '{ age: number; }' but required in type 'InputViaAssert'.ts(2741)

const input5: InputViaAssert = { age: 4, favNum: 7 }; // works fine
const input6: InputViaAssert = { age: 4, favNum: undefined }; // works fine
```

Basically, I'm wanting yup to create
```typescript
type IdealInput = {
  age: number
  favNum?: number | undefined
}
```
And here is what it is creating
```typescript
type idealInput = {
  age: number
  favNum: number | undefined // Note the lack of a "?".  favNum property MUST exist here
}
```

Im not sure if this is an intentional design decision, a bug, or just bad implementation of my part.  Any help is greatly appreciated

**Comment by jquense** [comment]
yeah at the moment this is neither a bug or a design choice, more a limitation...TS doesn't make it easy to conditionally add `?` modifiers over a mapped type

**Comment by MFogleman** [comment]
Thanks for the prompt reply, and the sanity check.  Should there maybe be a known issues reference or something for the project, to notify people of edge cases such as this?

**Comment by slavovojacek** [comment]
Not ideal but I suppose you could use the following:

```typescript
type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>

type IdealInput = PartialBy<InputViaTypeOf, 'favNum'>
```

See https://stackoverflow.com/a/54178819 for reference.

**Comment by slavovojacek** [comment]
@jquense @MFogleman I think [this suggestion](https://github.com/jquense/yup/issues/1210) might make it possible.

**Comment by s-abstract** [comment]
```TSX
/**
   * The type to make certain fields optional.
   * Supports OR, AND operators and all their combinations.
   *
   * @example
   * type Point3D = {x: number; y: number; z: number};
   *
   * Partial<Point3D, 'x' | 'y'> // To make optional one of x OR y
   * Partial<Point3D, ['x' | 'y']> // To make optional both x AND y
   * Partial<Point3D> // To make optional all, for backward compatibility with built in typescript Partial type
   */
  export type Optional<
    T extends Record<string, unknown>,
    U extends keyof T | (keyof T)[] = (keyof T)[]
  > = U extends (keyof T)[]
    ? Eval<
        {
          [key in U[number]]?: T[key];
        } &
          Omit<T, U[number]>
      >
    : U extends keyof T
    ? Eval<
        {
          [key in U]?: T[key];
        } &
          Omit<T, U>
      >
    : never;
```

I have a set of my own types like this Optional one. It works better than Partial in a way that it preserves dynamic types logic and not flattening it. I use it to solve this issue may be it will be helpful for someone.

**Comment by OoDeLally** [comment]
https://github.com/jquense/yup/issues/1210#issuecomment-1049817680
